hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java

```diff
@@ -222,11 +250,55 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
       throw new RuntimeException("Interrupted during commit", e);
 
     } finally {
-      cleanupMetadataAndUnlock(threw, newMetadataLocation, lockId);
+      cleanupMetadataAndUnlock(commitStatus, newMetadataLocation, lockId);
+    }
+  }
+
+  /**
+   * Attempt to load the table and see if any current or past metadata location matches the one we were attempting
+   * to set. This is used as a last resort when we are dealing with exceptions that may indicate the commit has
+   * failed but are not proof that this is the case. Past locations must also be searched on the chance that a second
+   * committer was able to successfully commit on top of our commit.
+   *
+   * @param newMetadataLocation the path of the new commit file
+   * @return Commit Status of Success, Failure or Unknown
+   */
+  private CommitStatus checkCommitStatus(String newMetadataLocation, TableMetadata metadata) {
+    int maxAttempts = PropertyUtil.propertyAsInt(metadata.properties(), COMMIT_NUM_STATUS_CHECKS,
+        COMMIT_NUM_STATUS_CHECKS_DEFAULT);
+    return innerCommitCheck(newMetadataLocation, 1, maxAttempts);
+  }
+
+  private CommitStatus innerCommitCheck(String newMetadataLocation, int attempt, int maxAttempts) {
+    try {
+      Thread.sleep(1000);
+      TableMetadata metadata = refresh();
+      String metadataLocation = metadata.metadataFileLocation();
+      boolean commitSuccess = metadataLocation.equals(newMetadataLocation) ||
+          metadata.previousFiles().stream().anyMatch(log -> log.file().equals(newMetadataLocation));
+      if (commitSuccess) {
+        LOG.info("Commit status check: Commit to {}.{} of {} succeeded", database, tableName, newMetadataLocation);
+        return CommitStatus.SUCCESS;
+      } else {
+        LOG.info("Commit status check: Commit to {}.{} of {} failed", database, tableName, newMetadataLocation);
+        return CommitStatus.FAILURE;
+      }
+    } catch (Throwable checkFailure) {
+      if (attempt < maxAttempts) {
+        LOG.error("Cannot check if commit to {}.{} exists. Retry attempt {} of {}.",
+            database, tableName, attempt, maxAttempts, checkFailure);
+        return innerCommitCheck(newMetadataLocation, attempt + 1, maxAttempts);
```

Comment:
Nit: wouldn't it be easier to read with a loop instead of recursion?

Reply:
The for loop would need early returns which i'm not a fan of, but I know not everyone has my taste. I'll switch it out if that makes it clearer

Reply:
I think it depends on personal style what we find easier to understand. As mentioned below, no strong feelings just an idea, so as the one writing the code it is really up to you 😃